### PR TITLE
Ldap fix

### PR DESCRIPTION
**Pull request by anusudarsan**
@losipiuk @amrutagokhale can you take a look?

### CODE REVIEW

File: tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java [hunk 1]
```diff
@@ -99,12 +102,22 @@ private String addLdapDefinition(LdapObjectDefinition ldapObjectDefinition)
 
         ldapObjectDefinition.getAttributes()
                 .forEach((k, v) -> entries.put(new BasicAttribute(k, v)));
+
+        List<ModificationItem> modificationItems = newArrayList();
+        for (String key: ldapObjectDefinition.getModificationAttributes().keySet())
```
Comment (amrutagokhale) [inline tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java]: Is it possible to get rid of this `for` clause altogether and convert this too to streams API method call?

**Comment by amrutagokhale** [comment]
Looks good % one comment.

File: tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java (hunk 1 above)
Comment (anusudarsan) [inline tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java]: done

File: README.md [hunk 2]
```diff
@@ -470,9 +470,17 @@ These definitions have objects analogous to an `OpenLDAP` entry definition and s
                 LdapObjectDefinition.builder("TestUser")
                         .setDistinguishedName("uid=testuser,ou=Test,dc=tempto,dc=com")
                         .setAttributes(testUserAttributes())
+                        .addModificationAttributes(memberOfAttributes())
                         .addObjectClasses(Arrays.asList("person", "inetOrgPerson"))
                                 .build();
 
+        private static Map<String, List<String>> memberOfAttributes()
+        {
+                Map<String, List<String>> attributes = newHashMap();
```
Comment (losipiuk) [inline README.md]: As a rule of thumb if `Map<>` is returned we should use `ImmutableMap` - not a HashMap.

File: tempto-ldap/src/main/java/com/teradata/tempto/fulfillment/ldap/LdapObjectDefinition.java [hunk 3]
```diff
@@ -28,12 +29,14 @@
     private final String distinguishedName;
     private final Map<String, String> attributes;
     private final List<String> objectClasses;
+    private final Map<String, List<String>> modificationAttributes;
 
-    private LdapObjectDefinition(String id, String distinguishedName, Map<String, String> attributes, List<String> objectClasses)
+    private LdapObjectDefinition(String id, String distinguishedName, Map<String, String> attributes, Map<String, List<String>> modificationAttributes, List<String> objectClasses)
```
Comment (losipiuk) [inline tempto-ldap/src/main/java/com/teradata/tempto/fulfillment/ldap/LdapObjectDefinition.java]: Maybe MultiMap would be better?
Do we care about order of values?

File: tempto-ldap/src/main/java/com/teradata/tempto/fulfillment/ldap/OpenLdapObjectDefinitions.java [hunk 4]
```diff
@@ -47,9 +48,17 @@
             LdapObjectDefinition.builder("TestUser")
                     .setDistinguishedName("uid=testuser,ou=Test,dc=tempto,dc=com")
                     .setAttributes(testUserAttributes())
+                    .addModificationAttributes(memberOf())
                     .addObjectClasses(Arrays.asList("person", "inetOrgPerson"))
                             .build();
 
+    private static Map<String, List<String>> memberOf()
```
Comment (losipiuk) [inline tempto-ldap/src/main/java/com/teradata/tempto/fulfillment/ldap/OpenLdapObjectDefinitions.java]: rename to `memberOfAttributes` or `buildMemberOfAttributes`. Or just inline in TestUser definition.

File: tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java [hunk 5]
```diff
@@ -99,12 +101,22 @@ private String addLdapDefinition(LdapObjectDefinition ldapObjectDefinition)
 
         ldapObjectDefinition.getAttributes()
                 .forEach((k, v) -> entries.put(new BasicAttribute(k, v)));
+
+        List<ModificationItem> modificationItems = newArrayList();
+        ldapObjectDefinition.getModificationAttributes().keySet()
+                .forEach(key -> modificationItems.addAll(ldapObjectDefinition.getModificationAttributes().get(key)
+                                .stream()
+                                .map(value -> new ModificationItem(DirContext.ADD_ATTRIBUTE, new BasicAttribute(key, value)))
+                                .collect(Collectors.toList())));
```
Comment (losipiuk) [inline tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java]: IMO this shape is nicer:
```
        List<ModificationItem> modificationItems = ldapObjectDefinition.getModificationAttributes().entrySet().stream()
                .flatMap(entry -> entry.getValue().stream()
                        .map(attribute -> new ModificationItem(DirContext.ADD_ATTRIBUTE, new BasicAttribute(entry.getKey(), attribute))))
                .collect(Collectors.toList());
```

File: tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java [hunk 6]
```diff
@@ -129,15 +124,17 @@ private String addLdapDefinition(LdapObjectDefinition ldapObjectDefinition)
...
-    public List<String> addLdapDefinitions(List<LdapObjectDefinition> ldapObjectDefinitions)
+    public void addLdapDefinitions(List<LdapObjectDefinition> ldapObjectDefinitions)
     {
+        DirContext context = createContext(ldapUrl, ldapAdminDistinguishedName, ldapAdminPassword);
         try {
-            return ldapObjectDefinitions.stream()
-                    .map(this::addLdapDefinition)
-                    .collect(toList());
+            for (LdapObjectDefinition ldapObjectDefinition: ldapObjectDefinitions)
+            {
```
Comment (losipiuk) [inline tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java]: put brace in previous line

File: tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java [hunk 7]
```diff
@@ -129,15 +124,17 @@ private String addLdapDefinition(LdapObjectDefinition ldapObjectDefinition)
...
+                addLdapDefinition(ldapObjectDefinition, context);
+            }
         }
         finally {
-            closeQuietly();
```
Comment (losipiuk) [inline tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java]: Oh - so it was broken before as `addLdapDefinitions` could only be called once. And on second call `context` would have already been closed.

File: tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java [hunk 8]
```diff
@@ -129,15 +124,17 @@ private String addLdapDefinition(LdapObjectDefinition ldapObjectDefinition)
...
     @Override
-    public List<String> addLdapDefinitions(List<LdapObjectDefinition> ldapObjectDefinitions)
+    public void addLdapDefinitions(List<LdapObjectDefinition> ldapObjectDefinitions)
     {
+        DirContext context = createContext(ldapUrl, ldapAdminDistinguishedName, ldapAdminPassword);
         try {
-            return ldapObjectDefinitions.stream()
-                    .map(this::addLdapDefinition)
-                    .collect(toList());
+            for (LdapObjectDefinition ldapObjectDefinition: ldapObjectDefinitions)
```
Comment (ArturGajowy) [inline tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java]: missing space before ':'

File: tempto-ldap/src/main/java/com/teradata/tempto/fulfillment/ldap/LdapObjectDefinition.java (hunk 3 above)
Comment (ArturGajowy) [inline tempto-ldap/src/main/java/com/teradata/tempto/fulfillment/ldap/LdapObjectDefinition.java]: I'd stick with `Map<..., List<...>>` in public API. Users can always create a multimap and do `Multimaps.asMap` if they prefer.

File: README.md (hunk 2 above)
Comment (anusudarsan) [inline README.md]: done

File: tempto-ldap/src/main/java/com/teradata/tempto/fulfillment/ldap/LdapObjectDefinition.java (hunk 3 above)
Comment (anusudarsan) [inline tempto-ldap/src/main/java/com/teradata/tempto/fulfillment/ldap/LdapObjectDefinition.java]: Also, I wanted to keep the values in the order.

File: tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java (hunk 5 above)
Comment (anusudarsan) [inline tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java]: done. thanks

File: tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java (hunk 7 above)
Comment (anusudarsan) [inline tempto-ldap/src/main/java/com/teradata/tempto/internal/fulfillment/ldap/DefaultLdapObjectEntryManager.java]: yes.

**Comment by anusudarsan** [comment]
addressed comments.

File: tempto-ldap/src/main/java/com/teradata/tempto/fulfillment/ldap/LdapObjectDefinition.java [hunk 9]
```diff
@@ -92,6 +101,12 @@ public LdapObjectDefinitionBuilder setAttributes(Map<String, String> attributes)
             return this;
         }
 
+        public LdapObjectDefinitionBuilder addModificationAttributes(Map<String, List<String>> modificationAttributes)
```
Comment (losipiuk) [inline tempto-ldap/src/main/java/com/teradata/tempto/fulfillment/ldap/LdapObjectDefinition.java]: Actually better name would be `setModificationAttributes`. Add suggest that you are merging previous value with one passed as parameter.
Same applies to `addObjectClassess` below. You can change that also if that is not huge pain.

**Comment by losipiuk** [comment]
Looks good modulo last comment.  Feel free to push when you address that one.

**Pull request merged**